README.md
![Flutjs Logo](./Logo.svg)

# Welcome to Flutjs project 😀

Flutjs is a javascript framework for creating beautiful, fast and lightweight websites.
As the name suggests, Flutejs is based on Flutter for its simplicity and easy learning curve and the ability to easily extend.

## Getting Stated:

To get start with Flutjs just follow the steps below:

### With boilerplate

Just go to [Flutjs Template Ts](https://github.com/LUK3D/flutjs-template-ts) repository and click on "Use This repository" and you are good to go.


### With Nodejs:

1 - Install the latest version of Nodejs.

2 - Start new nodejs project

  `npm init`

3 - Install Flutjs with npm or yarn:

>npm

`npm i @luk3d/flutjs`
>yarn

`yarn add @luk3d/flutjs`

Congrats!! 🎉 You have Installed Flujs successfully on your project. Now jump to hello world section the create a simple Flutjs website/webApp.


<img src="./docs/screenshots/flutjsCounterApp.png" width="500px">




### With Vanilla JS:


Add this link to the header tag of your html page:
```html
<script type="moduel" src="https://cdn.jsdelivr.net/npm/@luk3d/flutjs@0.2.0/dist/flutjs.min.js"></script>
```
>NOTE: you need to check the latest vesio here:
[Flutjs Latest](https://www.jsdelivr.com/package/npm/@luk3d/flutjs?path=dist)


## Simple Counter App: ⚡

```js
/**Basic Flutjs App With Tailwind */
// Stateful Variable 
var counterState = new StateManager();
//Defining an Observable State
counterState.obs(new State({ key: "counter", val: 0 }));

WindiApp({
  title: "Flutjs test",

  home: Scaffold({
    appBar: AppBar({
      backgroundColor: Colors.Blue[500],
      foregroundColor: Colors.Gray[100],
      title: Text("Welcome to Flutjs"),
      elevation: 2,
    }),
    body: Stack({
      children: [
        Container({
          color: Colors.Gray[50],
          width: 100,
          height: 100,
          child: Expanded({
            child: Column({
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                //Binding The widget to the State
                counterState.Bind(
                  Container({
                    key: "counter",
                    child: Text("{counter}", {
                      textStyle: TextStyle({
                        fontWeight: FontWeight.w700,
                        fontSize: 40,
                        color: Colors.Gray[500],
                      }),
                    }),
                  })
                ),
              ],
            }),
          }),
        }),
        Container({
          key: "addButton",
          child: Positioned({
            bottom: 20,
            right: 20,
            child: Container({
              padding: EdgeInsets.symmetric({ horizontal: 13, vertical: 20 }),
              decoration: BoxDecoration({
                color: Colors.Sky[500],
                borderRadius: BorderRadius.all(Radius.circular(100)),
              }),
              child: Text("+", {
                textStyle: TextStyle({
                  fontWeight: FontWeight.w600,
                  fontSize: 20,
                  color: Colors.White,
                }),
              }),
            }),
          }),
        }),
      ],
    }),
  }),
  // Executed when the document finish rendering the page
  mounted: () => {
    var i = 0;
    counterState.setState(`0`);
    document
      .querySelector('[key="addButton"]')
      ?.addEventListener("click", () => {
        //Updating the value of state
        counterState.setState(`${i++}`);
      });
  },
});

```

> NOTE: Just like in Flutter, every component is a widget...



## UPDATES TABLE 🚀

Date | Updates|
-|-|
 19-02-2022| `Text Widget` now supports `textStyle` property. Once compiled all values added on the textStyle attr will be translated to css attr of the text element |
 16-02-2022| The `App Widget` now fills the width and height of the screen by using 100% of `vw` and `vh` `css` units |
 16-02-2022| The `Extended Widget` now replaces himself with his child once compiled |
 16-02-2022| `Width` and `Height`  Where added on the widget constructor. |
 16-02-2022| Html output improvement |

  

## Under Development:🚧

- A Set of widgets like Material Design components.
- Route management
- State management
- Decorations (TextStyle, BoxDecoration etx...)
- Icons Library..
- Routing
- Custom cli tool for fast project creation.
- VsCode Extension for auto completion and snippets for fast coding...


CHANGELOG.md
# v0.2.4-alpha-1 (2022-06-15)

This version comes with Some Features, Updates and Enhancements
## Feature
* 🚀 State Management  - Now we have a basic state Management.
* Added `Image` Widget.
* Added `Painting` Class with `BlendMode` list.
* Added `BoxFit` Class with `BoxFit` list.
* Added `BoxShadow` Class.
* Now we have the  (flut-hero)[https://www.npmjs.com/package/@luk3d/flut-hero] package for icons
* Added `Offset` Class.

## Enhancements
* Added `BoxShadow` List To `BoxDecoration`.
* Added `BoxDecoration` To `Button` Widget.
## Bug Fixes




# v0.2.3-alpha-1 (2022-06-13)
This version comes with Some Features, Updates and Enhancements
## Feature
* Added `Radius` Class.
* Added `BorderRadiusGeometry` Class.
* Added `BorderRadius` property to the Container Component.
* Added `Spacer Widget`.
* Added `Cssprops` with `CssProperties` class that maps the Flutjs properties to css attributes.
* Added `Stack Widget`.
* Added `Positioned`.

## Enhancements
* Removed Tailwind Css Dependency.

## Bug Fixes
* Container now have Width and Hight of Auto as default value.
* Column widget now have the correct MainAxisDirection.
* Paddings now works well
* Added Margin and Padding Attributrs to the container



# v0.2.1-alpha-1 (2022-06-11)
This version adds new Features and some Bug Fixes.
## Feature
* Added AbsoluteUnits and RelativeUnits Types to define the length on all components that requires number and unit.
* Added `width_size_measurement_unit<AbsoluteUnits|RelativeUnits>` and `height_size_measurement_unit<AbsoluteUnits|RelativeUnits>` to the Container 

```ts
 Container(
    {
        color:Colors.White,
        height:40,
        width:70,
        width_size_measurement_unit:AbsoluteUnits.pixels,
        height_size_measurement_unit:AbsoluteUnits.pixels,
    }
)
```


Component to define the measurement unit for width and height.
* Added White and Black Colors.

## Bug Fixes
* Colors Class now return the correct colors from a given hexadecimal value. 

src/example.ts
/**Basic Flutjs App With Tailwind */
import {
  Spacer,
  Expanded,
  Row,
  RelativeUnits,
  BoxDecoration,
  Padding,
  Column,
  Container,
  Text,
  Colors,
  EdgeInsets,
  MainAxisAlignment,
  CrossAxisAlignment,
  TextStyle,
  Radius,
  AbsoluteUnits,
} from "./index.js";
import { BorderRadius } from "./painting/BorderRadiusGeometry.js";
import { BoxShadow } from "./painting/boxShadow.js";
import { State, StateManager } from "./state/stateManager.js";
import { Offset } from "./utils/offset.js";
import { Button } from "./widgets/Button.js";
import { Image } from "./widgets/image.js";
import { Positioned } from "./widgets/positioned.js";
import { Stack } from "./widgets/stack.js";
import { Scaffold, FlutApp, AppBar } from "./flut/index.js";

var counter = new StateManager(new State({key:"counter", val:1}));

FlutApp({
  title: "Flutjs test",

  home: Scaffold({
    appBar: AppBar({
      backgroundColor: Colors.White,
      foregroundColor: Colors.Gray[500],
      title: Text("Welcome to Flutjs"),
      elevation: 3,
      
    }),
    body: Container({
        color: Colors.Gray[50],
        width:100,
        height:100,
        child:
        counter.Bind( Expanded({
          child:  Column({
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
            
              Text("Hello World {counter} times!"), 
              
                // @ts-ignore
                ("{counter}" == '5')?
                   Text("🤩 {counter}", {textStyle: TextStyle({ fontSize: 20 })}):
                   Text("😩 {counter}", {textStyle: TextStyle({ fontSize: 20 })}),

              Button({
                padding:EdgeInsets.only({top:10,bottom:10, right:15,left:15}),
                label: Text("Clica-Me"), onPressed: () => {
                  counter.setState((counter.state?.val as number)+1)
                 },
                 boxDecoration:BoxDecoration({
                  boxShadow:[
                    new BoxShadow({
                      color:Colors.Gray[300],
                      blurRadius:10,
                      spreadRadius:10,
                      offset:Offset(0,10)
                      
                    }),
                    new BoxShadow({
                      color:Colors.Gray[100],
                      blurRadius:10,
                      spreadRadius:10,
                      offset:Offset(0,15)
                      
                    }),
                  ]
                 })
              }),

            ],
        }),
        }))
    }),
  }),

});


